File: src/main/java/komu/blunt/analyzer/IdentifierMapping.java
package komu.blunt.analyzer;

import komu.blunt.objects.Symbol;

import java.util.HashMap;

final class IdentifierMapping {
    
    private final IdentifierMapping parent;
    private final HashMap<Symbol, Symbol> mappings = new HashMap<Symbol, Symbol>();
    
    public IdentifierMapping() {
        this.parent = null;    
    }
    
    private IdentifierMapping(IdentifierMapping parent) {
        this.parent = parent;
    }
    
    public Symbol get(Symbol var) {
        for (IdentifierMapping mapping = this; mapping != null; mapping = mapping.parent) {
            Symbol sym = mapping.mappings.get(var);
            if (sym != null)
                return sym;
        }

        return var;
    }

    public void put(Symbol oldName, Symbol newName) {
        Symbol old = mappings.put(oldName, newName);
        if (old != null)
            throw new IllegalArgumentException("duplicate mapping for: " + oldName);
    }
    
    public IdentifierMapping extend() {
        return new IdentifierMapping(this);
    }
}

File: src/main/kotlin/komu/blunt/analyzer/IdentifierMapping.kt
package komu.blunt.analyzer

import komu.blunt.objects.Symbol

class IdentifierMapping(val parent: IdentifierMapping? = null) {

    private val mappings = hashMap<Symbol,Symbol>()
    private val sequence: Sequence = if (parent != null) parent.seq() else Sequence()

    private fun seq() = sequence

    fun get(v: Symbol): Symbol {
        var mapping = this

        while (true) {
            val sym = mapping.mappings[v]
            if (sym != null)
                return sym

            val parent = mapping.parent
            if (parent == null)
                break
            else
                mapping = parent
        }

        return v
    }

    fun set(oldName: Symbol, newName: Symbol) {
        val old = mappings.put(oldName, newName)
        if (old != null)
            throw IllegalArgumentException("duplicate mapping for '$oldName'");
    }

    fun extend() = IdentifierMapping(this)

    /**
     * Creates a new unique symbol based on name, installs it on the mapping and returns it.
     */
    fun freshMappingFor(name: Symbol): Symbol {
        val v = Symbol("\$${name}_${sequence.next()}")
        this[name] = v
        return v
    }
}

class Sequence {
    private var num = 1

    fun next() = num++
}
